retis/src/collect/collector.rs

```diff
@@ -513,7 +526,12 @@ mod tests {
         fn register_cli(&self, cli: &mut DynamicCommand) -> Result<()> {
             cli.register_module_noargs(SectionId::Skb)
         }
-        fn init(&mut self, _: &CliConfig, _: &mut ProbeBuilderManager) -> Result<()> {
+        fn init(
+            &mut self,
+            _: &CliConfig,
+            _: &mut ProbeBuilderManager,
+            _: &mut Event,
```

Comment:
Collectors do not generate events, but event sections. Then can generate 0:n sections and put them into a single event. Also the initialization is done at a single point in time so there's no real need to have more than one event IMO, as all sections can be part of the same. Or do you have a use case in mind?

Reply:
I'm not sure I have a particular use case in mind. But there is something that does not quite fit.

Normal events (from eBPF) have different event sections. These sections are unrelated to each other (modular) but there is something that ties them together (into an event): they all refer to the same packet.

With metadata events we are adding 100% unrelated data into the same event which feels kinda weird, specially if the event itself has no semantics at all that ties the sections together (the "packet" for ebpf events).

IIUC, the idea is for modules to also generate md events during collection. When that happens, will we collect all event_sections into a single event? or will we have events with a single section? I guess the later, so this makes squeezing all event sections into a single event a bit weird.

Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That  would "tie" all sections together (they are each module's initial config) and would not seem so loose.

An alternative is to have each module generate a full event.

Reply:
> Normal events (from eBPF) have different event sections. These sections are unrelated to each other (modular) but there is something that ties them together (into an event): they all refer to the same packet.
> 
> With metadata events we are adding 100% unrelated data into the same event which feels kinda weird, specially if the event itself has no semantics at all that ties the sections together (the "packet" for ebpf events).
> 
> IIUC, the idea is for modules to also generate md events during collection. When that happens, will we collect all event_sections into a single event? or will we have events with a single section? I guess the later, so this makes squeezing all event sections into a single event a bit weird.

Discussion is a bit mangled between here and the other thread, please see my answer in there.

> Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That would "tie" all sections together (they are each module's initial config) and would not seem so loose.

The event being an "initial configuration" one comes from the fact that `CommonEventMd` is present. That is vary much like what's make a BPF event: the section about the probe is there. (In addition both have a `CommonEvent` section).

> An alternative is to have each module generate a full event.

I'm not 100% against that TBF.

Reply:
> > Normal events (from eBPF) have different event sections. These sections are unrelated to each other (modular) but there is something that ties them together (into an event): they all refer to the same packet.
> > With metadata events we are adding 100% unrelated data into the same event which feels kinda weird, specially if the event itself has no semantics at all that ties the sections together (the "packet" for ebpf events).
> > IIUC, the idea is for modules to also generate md events during collection. When that happens, will we collect all event_sections into a single event? or will we have events with a single section? I guess the later, so this makes squeezing all event sections into a single event a bit weird.
> 
> Discussion is a bit interleaved between here and the other thread, please see my answer in there.
> 

Sorry for the mess.

> > Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That would "tie" all sections together (they are each module's initial config) and would not seem so loose.
> 
> The event being an "initial configuration" one comes from the fact that `CommonEventMd` is present. That is vary much like what's make a BPF event: the section about the probe is there. (In addition both have a `CommonEvent` section).
> 

I agree that we should keep events loosely defined (see other thread).

So, let's see what kind of events we can have:

**A** - Events with probe section: What all the sections have in common is that they somehow describe the event that was originated by such probe. These event sections that come from the BPF side but also metadata events originated by, for instance, post-processing tools that provide extra data to the event.

**B** - Events without probe section: What these events have in common is the timestamp, (part of the `CommonEvent` section).

For events of type B, one could argue that, if some sections take some time to generate, the timestamp will be more accurate for some sections than others. So I see this "connection" between sections kind of loose (or at least more loose than for events with a probe section). We could add a section (the equivalent for the probe section for these events) that defines the "stage" this event was generated at, (e.g: init, end, ... ). However, this will make spontaneous events generated during collection (e.g: due to some config change) also kind of loose and only bound by the timestamp. So I guess I'll have to live with this as I don't have a better solution.

Reply:
> > Discussion is a bit interleaved between here and the other thread, please see my answer in there.
> 
> Sorry for the mess.

No worries, this is only because of the GH interface :)

> > > Have you considered adding something that indicates that the first event is that, an "initial-configuration" event? That would "tie" all sections together (they are each module's initial config) and would not seem so loose.
> > 
> > The event being an "initial configuration" one comes from the fact that `CommonEventMd` is present. That is vary much like what's make a BPF event: the section about the probe is there. (In addition both have a `CommonEvent` section).
> 
> I agree that we should keep events loosely defined (see other thread).
> 
> So, let's see what kind of events we can have:
> 
> **A** - Events with probe section: What all the sections have in common is that they somehow describe the event that was originated by such probe. These event sections that come from the BPF side but also metadata events originated by, for instance, post-processing tools that provide extra data to the event.
> 
> **B** - Events without probe section: What these events have in common is the timestamp, (part of the `CommonEvent` section).

And that's only for now. I would just summarize events as being a collection of sections sharing something in common. Be it a probe + ts or something else. There's no real concept of "metadata" in reality (I'm trying to stop using the word).

> For events of type B, one could argue that, if some sections take some time to generate, the timestamp will be more accurate for some sections than others. So I see this "connection" between sections kind of loose (or at least more loose than for events with a probe section). We could add a section (the equivalent for the probe section for these events) that defines the "stage" this event was generated at, (e.g: init, end, ... ). However, this will make spontaneous events generated during collection (e.g: due to some config change) also kind of loose and only bound by the timestamp. So I guess I'll have to live with this as I don't have a better solution.

I do agree that "config" sections only sharing a timestamp is the one case not perfectly fitting. We can still issue a single event per "logical data", aka. allowing modules to issue a full event and not a section only at init time. This would also map better to what could happen later: modules issuing an update event sections for some kind of config. A full event would probably be generated in such case.

Reply:
> I do agree that "config" sections only sharing a timestamp is the one case not perfectly fitting. We can still issue a single event per "logical data", aka. allowing modules to issue a full event and not a section only at init time. This would also map better to what could happen later: modules issuing an update event sections for some kind of config. A full event would probably be generated in such case.

I guess so, yes. That's why I originally suggested it. Either way is fine, specially since we don't have per-module events and I cannot see a practical example.